rangeproof/proof.go

```diff
@@ -106,11 +107,12 @@ Notes:
 */
 
 type (
-	// Statement states that an attribute m satisfies Factor*m-Bound >= 0, and that Factor*m-Bound
-	// can be split into squares with the given Splitter. E.g. if Factor = 1 then m >= k. Defaults to
-	// four square splitter when splitter is not specified.
+	// Statement states that an attribute m satisfies Sign*(Factor*m-Bound) >= 0, and that
+	// Sign*(Factor*m-Bound) can be split into squares with the given Splitter. E.g. if Factor = 1
+	// then Factor*m >= Bound. Defaults to four square splitter when splitter is not specified.
 	Statement struct {
-		Factor   int
+		Sign     int
```

Comment:
If Sign can only be 1 or -1 I would expect this not to be of type `int` but of type `boolean` (or custom type)

Reply:
Indeed it would make sense to enforce correctness by using a type that can assume only two values. There is however only one such type in Go: the `bool`. That means we have to assign `true` and `false` to `>=` and `<=`, while mathematically, those two are completely symmetrical: certainly one is not more true than the other. Superficially it might make sense to assign `true` to `>=` because in that case `sign` is `1` in the current implementation, but that is really only a consequence of the essentially arbitrary decision that was made to have the code prove `a*m - k >= 0`, as opposed to `k - a*m >= 0`. In that case, everything would have been reversed.

So we have to choose one of the following:

* enforce correctness by using a `bool`
* keep symmetry in the code between `>=` and `<=`. In addition, having `sign` a number equal to `1` or `-1` makes the code be closer to the mathematical meaning that it implements.

Given that there are [other ways](https://github.com/privacybydesign/gabi/pull/20/files#diff-cf05d5d6d185ced7f2779d054485f3747690c322784f484e932713174d193ec4R209) to enforce correctness, the latter approach makes more sense to me.